### PR TITLE
Structure tests

### LINKED ISSUE #215: Breakup tensornetwork_test?
It's getting pretty large, >1,600 lines of code.

It doesn't matter that much, and I'm not sure how we would break it apart. `tensornetwork_test_2`?

### PR DESCRIPTION
**Pull request by MichaelMarien**
Fixes #215. Split up in a
* `network_test.py`: contains unit tests for all methods of `TensorNetwork` class. Some to-do's are added for missing tests, for which - if agreed by reviewers - I can later make an Issue.
* `split_node_test.py`: all tests related to node split methods
* `serialization.py`: all tests related to saving and loading tensornetworks
* `tensornetwork_test.py`: All remaining tests, most tests that do slightly more than a pure minimal unites, although sometimes the boundary is a bit arbitrary.

No tests were removed in this PR, only moved and some extras added (where I noticed missing tests and could immediately add them myself).

### CODE REVIEW

**Comment by MichaelMarien** [comment]
@Thenerdstation , build is failing python3.5, not for 3.6 and 3.7:

FAILED: /home/travis/build/MichaelMarien/TensorNetwork/.pytype/pyi/tensornetwork/tests/network_test.pyi 
349pytype-single --imports_info /home/travis/build/MichaelMarien/TensorNetwork/.pytype/imports/tensornetwork.tests.network_test.imports --module-name tensornetwork.tests.network_test -V 3.5 -o /home/travis/build/MichaelMarien/TensorNetwork/.pytype/pyi/tensornetwork/tests/network_test.pyi --analyze-annotated --nofail --quick /home/travis/build/MichaelMarien/TensorNetwork/tensornetwork/tests/network_test.py
350File "/home/travis/build/MichaelMarien/TensorNetwork/tensornetwork/tests/network_test.py", line 22, in <module>: Can't find module 'jax.config'. [import-error]
...
...
E   ValueError: jaxlib is version 0.1.23, but this version of jax requires version 0.1.26.

Any idea how to fix this? Is newest jax no longer compatible with python3.5?

**Comment by Thenerdstation** [comment]
It should be. 

Try bumping the `jaxlib` version in `requirements_travis.txt`

**Comment by Thenerdstation** [comment]
Oh yeah, we aren't setting the version for Jax
https://github.com/google/TensorNetwork/blob/af16c219e3cc1a411c89314648eb059000c378b8/requirements_travis.txt#L2

Feel free to fix that in this PR.

**Comment by MichaelMarien** [comment]
> Oh yeah, we aren't setting the version for Jax
> 
> https://github.com/google/TensorNetwork/blob/af16c219e3cc1a411c89314648eb059000c378b8/requirements_travis.txt#L2
> 
> Feel free to fix that in this PR.

Seems jaxlib for python3.5 is not really keeping up
Collecting jaxlib>=0.1.27 (from -r requirements_travis.txt (line 2))
113  Could not find a version that satisfies the requirement jaxlib>=0.1.27 (from -r requirements_travis.txt (line 2)) (from versions: 0.1, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.8, 0.1.9, 0.1.11, 0.1.12, 0.1.13, 0.1.14, 0.1.15, 0.1.16, 0.1.17, 0.1.18, 0.1.19, 0.1.20, 0.1.21, 0.1.22, 0.1.23)
114No matching distribution found for jaxlib>=0.1.27 (from -r requirements_travis.txt (line 2))
115The command "pip install -r requirements_travis.txt" failed and exited with 1 during .

Do you want to freeze the Jax version to a lower one (latest compatible with jaxlib 0.1.23) or drop 3.5 dependency?

**Comment by Thenerdstation** [comment]
Looks like JAX is only aiming to support python 2.7 and 3.6

https://github.com/google/jax/blob/110634d50dab0820aa777e161325f479f64db865/.travis.yml#L8

We might be ok just supporting 3.6 and 3.7. I'm not sure there yes a strong argument in keeping 3.5 around

**Comment by MichaelMarien** [comment]
Don't think there is, 3.5 is getting really old itself.

**Comment by Thenerdstation** [comment]
Ok, go ahead and remove it from the Travis.

File: requirements_travis.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 jax
```
Comment (Thenerdstation) [inline requirements_travis.txt]: Add version for jax too.

File: tensornetwork/tests/network_test.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+
+def test_disconnect_dangling_edge_value_error(backend):
+  net = tensornetwork.TensorNetwork(backend=backend)
+  a = net.add_node(np.eye(2))
+  with pytest.raises(ValueError):
+    net.disconnect(a[0])
+
+
+def test_remove_trace_edge():
+  #todo
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: Why are these todos?

**Comment by MichaelMarien** [comment]
@Thenerdstation  > * `network_test.py`: contains unit tests for all methods of `TensorNetwork` class. Some to-do's are added for missing tests, for which - if agreed by reviewers - I can later make an Issue.

File: tensornetwork/tests/network_test.py (hunk 2 above)
Comment (MichaelMarien) [reply tensornetwork/tests/network_test.py]: @Thenerdstation  > * `network_test.py`: contains unit tests for all methods of `TensorNetwork` class. Some to-do's are added for missing tests, for which - if agreed by reviewers - I can later make an Issue.

Comment (MichaelMarien) [reply tensornetwork/tests/network_test.py]: Somehow my quoting of my previous message got messed up. @Thenerdstation, by splitting up the tests, I noticed some methods of TensorNetwork do not have the most basic of unit tests. Some of them are private, but I'm a bit of an annoying test guy and would like to see basic unittests for all of them :). My plan was to make an issue for this after this PR and leave it as a good first issue. If the todos bother you I can remove them.

**Comment by Thenerdstation** [comment]
I'd say remove them. Really unit tests should only test the publicly facing methods, and private methods are open the change at any time.

**Comment by MichaelMarien** [comment]
> I'd say remove them. Really unit tests should only test the publicly facing methods, and private methods are open the change at any time.

Not sure I 100% agree with that, but that's fine. I removed the todo's for the private methods and added some tests for the missing public methods.

**Pull request merged**